onnxruntime/core/providers/shared/node_unit/node_unit.cc

```diff
@@ -26,20 +102,77 @@ void NodeUnit::InitForNode() {
   const auto& output_defs = node_.OutputDefs();
   // The 1st step is to hookup the NodeUnit with the NNAPI builder interface
   // So we are not handling quantization here now
-  // TODO, enable quantization
-  // auto qlinear_type = GetQLinearOpType(node_);
-  // if (qlinear_type == QLinearOpType::Unknown) {
-  // Not a Qlinear op, add all inputs/outputs
-  auto add_all_io = [](std::vector<IODef>& defs,
-                       const ConstPointerContainer<std::vector<NodeArg*>>& node_defs) {
-    defs.reserve(node_defs.size());
-
-    for (const auto def : node_defs) {
-      defs.push_back(NodeUnit::IODef{*def, std::nullopt});
+  auto qlinear_type = GetQLinearOpType(node_);
+  if (qlinear_type == QLinearOpType::Unknown ||
+      IsVariadicQLinearOp(qlinear_type)) {  // TODO, add variadic support
+    // Not a Qlinear op, add all inputs / outputs
+    auto add_all_io = [](std::vector<NodeUnitIODef>& defs,
+                         const ConstPointerContainer<std::vector<NodeArg*>>& node_defs) {
+      defs.reserve(node_defs.size());
+
+      for (const auto def : node_defs) {
+        defs.push_back(NodeUnitIODef{*def, std::nullopt});
+      }
+    };
+    add_all_io(inputs_, input_defs);
+    add_all_io(outputs_, output_defs);
+  } else if (IsUnaryQLinearOp(qlinear_type)) {
+    // Unary QLinear Op has 5 inputs
+    // x, x_scale, x_zp, y_scale, y_zp (optional)
+    inputs_.push_back(NodeUnitIODef{
+        *input_defs[0],
+        NodeUnitIODef::QuantParam{*input_defs[1], input_defs[2]}});
+
+    outputs_.push_back(NodeUnitIODef{
+        *output_defs[0],
+        NodeUnitIODef::QuantParam{*input_defs[3],
+                                  input_defs.size() > 4
+                                      ? input_defs[4]
+                                      : nullptr}});
+  } else if (IsBinaryQLinearOp(qlinear_type)) {
+    // Binary QLinear Op has 9 inputs
+    // x1, x1_scale, x1_zp, x2/w, x2_scale, x2_zp, y_scale , y_zp, B
+    inputs_.push_back(NodeUnitIODef{
+        *input_defs[0],
+        NodeUnitIODef::QuantParam{*input_defs[1], input_defs[2]}});
+    inputs_.push_back(NodeUnitIODef{
+        *input_defs[3],
+        NodeUnitIODef::QuantParam{*input_defs[4], input_defs[5]}});
+
+    if (input_defs.size() == 9) {  // has Bias
+      inputs_.push_back(NodeUnitIODef{
+          *input_defs[8],
+          std::nullopt});  // for Bias the scale and zp are optional
```

Comment:
>for Bias the scale and zp are optional [](http://example.com/codeflow?start=30&length=38)

beginner question - comment says optional, where would they be if they were provided?

Reply:
See https://github.com/onnx/onnx/blob/main/docs/Operators.md#inputs-8---9
```
B (optional) : T4
Optional 1D bias to be added to the convolution, has size of M. Bias must be quantized using scale = x_scale * w_scale and zero_point = 0
```